pkg/util/servicecidr/servicecidr.go

```diff
@@ -2,132 +2,244 @@ package servicecidr
-		return FallbackCIDR, fmt.Sprintf("failed to detect service CIDR, will fallback to %s, however this is probably wrong, please make sure the host cluster service cidr and virtual cluster service cidr match. Error details: failed to find IPv4 service CIDR: %v ; or IPv6 service CIDR: %v", FallbackCIDR, ipv4Err, ipv6Err)
-	}
-	if ipv4Err != nil {
-		return ipv6CIDR, fmt.Sprintf("failed to find IPv4 service CIDR, will use IPv6 service CIDR. Error details: %v", ipv4Err)
-	}
-	if ipv6Err != nil {
-		return ipv4CIDR, fmt.Sprintf("failed to find IPv6 service CIDR, will use IPv4 service CIDR. Error details: %v", ipv6Err)
+	// check if we need to use the new service cidr detection
+	vClusterService, err := client.CoreV1().Services(vClusterNamespace).Get(ctx, vClusterServiceName, metav1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("failed to get vCluster service %s: %v", vClusterServiceName, err)
 	}
 
-	if client == nil {
-		panic("client is nil")
+	// check if we already found the service cidr
+	if vClusterService.Annotations[ServiceCIDRAnnotation] != "" {
+		klog.Infof("using cached service cidr from annotation: %s", vClusterService.Annotations[ServiceCIDRAnnotation])
+		return vClusterService.Annotations[ServiceCIDRAnnotation], nil
 	}
 
-	// Both IPv4 and IPv6 are configured, we need to find out which one is the default
-	policy := corev1.IPFamilyPolicyPreferDualStack
-	testService, err := client.CoreV1().Services(namespace).Create(ctx, &corev1.Service{
-		ObjectMeta: metav1.ObjectMeta{
-			GenerateName: "test-service-delete-me-",
-		},
-		Spec: corev1.ServiceSpec{
-			Ports: []corev1.ServicePort{
-				{
-					Port: 80,
+	// create function to check if the ip is in the service cidr
+	isInRange := func(ip net.IP) bool {
+		policy := corev1.IPFamilyPolicySingleStack
+		ipFamily := corev1.IPv4Protocol
+		if ip.To4() == nil {
+			ipFamily = corev1.IPv6Protocol
+		}
+
+		// create an actual service to check this
+		testService, err := client.CoreV1().Services(vClusterNamespace).Create(ctx, &corev1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: "test-service-delete-me-",
+			},
+			Spec: corev1.ServiceSpec{
+				Ports: []corev1.ServicePort{
+					{
+						Port: 80,
+					},
 				},
+				IPFamilyPolicy: &policy,
+				IPFamilies:     []corev1.IPFamily{ipFamily},
+				ClusterIP:      ip.String(),
 			},
-			IPFamilyPolicy: &policy,
-		},
-	}, metav1.CreateOptions{})
-
-	if err == nil {
-		defer func() {
-			_ = client.CoreV1().Services(namespace).Delete(ctx, testService.GetName(), metav1.DeleteOptions{})
-		}()
-
-		// check if this is dual stack, and which family is default
-		if len(testService.Spec.IPFamilies) > 0 {
-			if testService.Spec.IPFamilies[0] == corev1.IPv4Protocol {
-				// IPv4 is the default
-				return fmt.Sprintf("%s,%s", ipv4CIDR, ipv6CIDR), ""
+		}, metav1.CreateOptions{})
+		if err == nil {
+			err := client.CoreV1().Services(vClusterNamespace).Delete(ctx, testService.GetName(), metav1.DeleteOptions{})
```

Comment:
Do we want to `defer` the deletion of services? We wouldnt want any any dangling test services

Reply:
Whats the benefit of a defer here instead of deleting it directly? In either case the service gets deleted

Reply:
deferring will make sure , even if there are panics, the test service still will get cleaned up.